pkg/kconfig/minimize.go

```diff
@@ -4,79 +4,65 @@
-	// First, check the base config as is, it is the smallest we can possibly get.
-	if res, err := pred(base); err != nil {
-		return nil, err
-	} else if res {
-		dt.Log("base config crashes")
-		return base, nil
-	}
-	// Since base does not crash, full config is our best bet for now.
-	current := full.clone()
-	var suspects []string
-	// Take half of the diff between base and full, apply to base and test.
-	// If this candidate config crashes, we commit it as new full and repeat the process.
-	// If it does not crash, try another half.
-	// If the crash is caused by a single config, this algorithm is guaranteed to find it.
-	// If the crash is caused by multiple configs, this algorithm will most likely find them (along with some
-	// additional unrelated configs that happened to be in the same half). However, amount of unrelated configs
-	// can be quite large if we are unlucky. Also note that we sort configs so that related configs are most
-	// likely situated together.
-	// Numerous improvements are possible for this simple algorithm.
-	// 1. We could split the config onto 4 parts and try all pairs, this should find all pairs of configs reliably.
-	// 2. We could continue trying to reduce a part even if removing the whole part fails. I.e. we try to remove
-	//    a half and it fails, we can try to remove half of the half, maybe that will succeed.
-top:
-	for len(diff) >= 2 {
-		half := len(diff) / 2
-		for _, part := range [][]string{diff[:half], diff[half:]} {
-			dt.Log("trying half: %v", part)
-			closure := kconf.addDependencies(base, full, part)
-			candidate := base.clone()
-			// Always move all non-tristate configs from full to base as we don't minimize them.
-			for _, cfg := range other {
-				candidate.Set(cfg.Name, cfg.Value)
-			}
-			for _, cfg := range closure {
-				candidate.Set(cfg, Yes)
-			}
-			res, err := pred(candidate)
-			if err != nil {
-				return nil, err
-			}
-			if res {
-				dt.Log("half crashed")
-				diff = part
-				current = candidate
-				suspects = closure
-				continue top
-			}
+	maxSteps int, dt debugtracer.DebugTracer) (*ConfigFile, error) {
+	diff, other := kconf.missingLeafConfigs(base, full)
+	dt.Log("kconfig minimization: base=%v full=%v leaves diff=%v", len(base.Configs), len(full.Configs), len(diff))
+
+	diffToConfig := func(part []string) (*ConfigFile, []string) {
+		if len(part) == 0 {
+			// We're testing the baseline config only.
+			return base, nil
+		}
+		suspects := kconf.addDependencies(base, full, part)
+		candidate := base.clone()
+		// Always move all non-tristate configs from full to base as we don't minimize them.
+		for _, cfg := range other {
+			candidate.Set(cfg.Name, cfg.Value)
 		}
-		dt.Log("both halves did not crash")
-		break
+		for _, cfg := range suspects {
+			candidate.Set(cfg, Yes)
+		}
+		return candidate, suspects
 	}
+	step := 1
+	result, err := generic.MinimizeSlice(generic.MinimizeConfig[string]{
+		Pred: func(diffs []string) (bool, error) {
```

Comment:
Please move the predicate out of line. "Configs" are supposed to be declarative and readable, but it's hard to read with significnat code chunks.